### PR TITLE
check for non-defining uses of RPIT

### LINKED ISSUE #111935: RPIT allows defining use with invalid args
**Note**: All examples in this post was fixed in #112842. However the issue persists for lifetime arguments. See [this comment](https://github.com/rust-lang/rust/issues/111935#issuecomment-1770497883) for more.

---

The following compiles since v1.62:
```rust
fn foo<T>() -> impl Sized {
    let _: () = foo::<u8>();
}
```

While the equivalent TAIT version doesn't compile because `Opaque<u8> == u8` is not a valid defining use for `Opaque` (because it is ambiguous whether `Opaque<T> = T` or `Opaque<T> = u8`):
```rust
#![feature(type_alias_impl_trait)]

type Opaque<T> = impl Sized;
fn foo<T>() -> Opaque<T> {
    let _: () = foo::<u8>();
    //~^ ERROR expected generic type parameter, found `u8`
}
```

I believe that TAIT behavior is the correct one because it's maximally compatible with future changes. Keeping this behavior for RPIT can also result in some surprising errors:
```rust
fn foo<T>(val: T) -> impl Sized {
    let _: u8 = foo(0u8);
    val
    //~^ ERROR concrete type differs from previous defining opaque type use
}
```

This was stabilized (most likely unintentionally) in #94081. Cc @oli-obk.

See related code:
https://github.com/rust-lang/rust/blob/776f22292fb5c2cc45f029d4b754662120732a6a/compiler/rustc_borrowck/src/region_infer/opaque_types.rs#L335

### PR DESCRIPTION
**Pull request by lcnr**
This PR requires defining uses of RPIT and the async functions return type to use unique generic parameters as type and const arguments, (mostly) fixing #111935. This changes the following snippet to an error (it compiled since 1.62):
```rust
fn foo<T>() -> impl Sized {
    let _: () = foo::<u8>(); //~ ERROR non-defining use of `impl Sized`
}
```
Since 1.62 we only checked that the generic arguments of opaque types are unique parameters for TAIT and ignored RPITs, so this PR changes the behavior here to be consistent.

For defining uses which do not have unique params as arguments it is unclear how the hidden type should map to the generic params of the opaque. In the following snippet, should the hidden type of `foo<T>::opaque` be `T` or `u32`.
```rust
fn foo<T>() -> impl Sized {
    let _: u32 = foo::<u32>();
    foo::<T>()
}
```
There are no crater regressions caused by this change.

---

The same issue exists for lifetime arguments which is not fixed by this PR, currently resulting in an ICE in mir borrowck (I wasn't able to get an example which didn't ICE, it might be possible):
```rust
fn foo<'a: 'a>() -> impl Sized {
    let _: &'static () = foo::<'static>();
    //~^ ICE opaque type with non-universal region substs
    foo::<'a>()
}
```
Fixing this for lifetimes as well is blocked on https://github.com/rust-lang/rust/issues/113916. Due to this issue, functions returning an RPIT with lifetime parameters equal in the region constraint graph would always result in an error, resulting in breakage found via crater: https://github.com/rust-lang/rust/pull/112842#issuecomment-1610057887
```rust
trait Trait<'a, 'b> {}
impl Trait<'_, '_> for () {}

struct Type<'a>(&'a ());
impl<'a> Type<'a> {
    // `'b == 'a`
    fn do_stuff<'b: 'a>(&'b self) -> impl Trait<'a, 'b> {
        // This fails as long there is something in the body
        // which adds the outlives constraints to the constraint graph.
        //
        // This is the case for nested closures.
        (|| ())() 
    
    }
}
```

### CODE REVIEW

**Comment by oli-obk** [comment]
@craterbot check

**Comment by aliemjay** [comment]
continuing on from https://github.com/rust-lang/rust/issues/112841#issuecomment-1600627589

I think the right way to fix tha case of async fn and RRPIT is to allow equal lifetimes to appear in the substs if and only if these lifetimes are assumed equal by the opaque type's definition
```rust
trait Trait<'a, 'b> {}
impl Trait<'_, '_> for () {}

type Tait1<'a, 'b> = impl Trait<'a, 'b>;
fn define1<'a>() -> Tait1<'a, 'a> {} // ERROR!


type Tait2<'a:' 'b, 'b: 'a> = impl Trait<'a, 'b>;
fn define2<'a>() -> Tait2<'a, 'a> {} // OK!
```

Because the user can't reference `Tait<'a, 'b>` without `'a` and `'b` being equal, per the well-formedness rules.

**Comment by aliemjay** [comment]
@craterbot run mode=check-only p=1 crates=https://crater-reports.s3.amazonaws.com/pr-112842/retry-regressed-list.txt

**Comment by lcnr** [comment]
limited this check to only type and consts params for now, cc #113916

@bors try

**Comment by lcnr** [comment]
@craterbot check

**Comment by aliemjay** [comment]
> tada Experiment **`pr-112842-2`** is completed!
> bar_chart 16 regressed and 2 fixed (329222 total)

all spurious :tada:

File: compiler/rustc_borrowck/src/region_infer/opaque_types.rs [hunk 1]
```diff
@@ -356,39 +356,29 @@ fn check_opaque_type_parameter_valid(
...
     for (i, arg) in opaque_type_key.args.iter().enumerate() {
+        if let Err(guar) = arg.error_reported() {
+            return Err(guar);
+        }
+
         let arg_is_param = match arg.unpack() {
             GenericArgKind::Type(ty) => matches!(ty.kind(), ty::Param(_)),
             GenericArgKind::Lifetime(lt) => {
-                matches!(*lt, ty::ReEarlyBound(_) | ty::ReFree(_))
+                if is_ty_alias {
```
Comment (aliemjay) [inline compiler/rustc_borrowck/src/region_infer/opaque_types.rs:374]: super tiny nit: I think it's nicer if we pull the conditional to the match arm and have the FIXME comment on the match arm for more flat code.

**Review by aliemjay — APPROVED** [review]

**Comment by lcnr** [comment]
This PR requires defining uses of RPIT and the async functions return type to use unique generic parameters as type and const arguments, (mostly) fixing #111935. This changes the following snippet to an error (it compiled since 1.62):
```rust
fn foo<T>() -> impl Sized {
    let _: () = foo::<u8>(); //~ ERROR non-defining use of `impl Sized`
}
```
Since 1.62 we only checked that the generic arguments of opaque types are unique parameters for TAIT and ignored RPITs, so this PR changes the behavior here to be consistent.

For defining uses which do not have unique params as arguments it is unclear how the hidden type should map to the generic params of the opaque. In the following snippet, should then hidden type of `foo<T>::opaque` be `T` or `u32`.
```rust
fn foo<T>() -> impl Sized {
    let _: u32 = foo::<u32>();
    foo::<T>()
}
```
There are no crater regressions caused by this change.

---

The same issue exists for lifetime arguments which is not fixed by this PR, currently resulting in an ICE in mir borrowck (I wasn't able to get an example which didn't ICE, it might be possible):
```rust
fn foo<'a: 'a>() -> impl Sized {
    let _: &'static () = foo::<'static>();
    //~^ ICE opaque type with non-universal region substs
    foo::<'a>()
}
```
Fixing this for lifetimes as well is blocked on https://github.com/rust-lang/rust/issues/113916. Due to this issue, functions returning an RPIT with lifetime parameters equal in the region constraint graph would always result in an error, resulting in breakage found via crater: https://github.com/rust-lang/rust/pull/112842#issuecomment-1610057887
```rust
trait Trait<'a, 'b> {}
impl Trait<'_, '_> for () {}

struct Type<'a>(&'a ());
impl<'a> Type<'a> {
    // `'b == 'a`
    fn do_stuff<'b: 'a>(&'b self) -> impl Trait<'a, 'b> {
        // This fails as long there is something in the body
        // which adds the outlives constraints to the constraint graph.
        //
        // This is the case for nested closures.
        (|| ())() 
    
    }
}
```

@rfcbot fcp merge

**Comment by oli-obk** [comment]
r=me with @aliemjay's comment addressed

@rustbot author

**Pull request merged**